### DISCUSSION TITLE
Compiling portable 6.9.8-10-Q8 issues

### DISCUSSION THREAD

**Discussion by sergenikov** [ImageMagick/ImageMagick#2005] [General] [answered]
Hello!

I am compiling 6.9.8-10-Q8 and observing odd behaviour of my built version for Windows. My company needs this particular version because future versions have breaking changes for us. I follow instructions on [ImageMagic on Windows](https://github.com/ImageMagick/ImageMagick6-Windows) and was able to build portable binary that runs. However, my binary doesn't behave the same way as the official non-portable Q8 I dug up on FTPs. Below are some issues I found while building. I would really appreciate some help with this. Please let me know if I forgot some more details.

Some build notes:
- I use VS 2017
- I build MTD version, but have also worked with MT with the same results.
- Tried with and w/o OpenMP (I think this gets statically linked in a portable - it always shows up in the features too, selected in configure.exe or not)
- Tried with and w/o OpenCL
- I've never been able to make the setup.exe installer, so I made it work my getting needed DLLs manually. At the moment, they are: `ucrtbase.dll, ucrtbased.dll, vcomp140.dll, vcom140d.dll, vcruntime140_1d.dll, vcruntime140d.dll, msvcp140.dll, msvcp140d.dll`.

1. I found no way to build a portable binary without having to modify `magick-baseconfig.h` file after running `configure.exe`. To build a portable version, when running `configure.exe` (and by looking at the source code of it and debugging) one needs to unselect "Use registry settings or hard-coded paths to located installed components". That will then result in `//#define MAGICKCORE_INSTALLED_SUPPORT` written into the baseconfig file. That's good, but I found that if I also don't comment out "zero configuration" line `//#define MAGICKCORE_ZERO_CONFIGURATION_SUPPORT 0`, I keep getting IM portable that uses zero configuration, so can't load any of my configs. Simply having it set to 0 seems to build it in zero-configuration mode regardless. Only making sure `MAGICKCORE_ZERO_CONFIGURATION_SUPPORT ` is completely undefined I am able to get a portable build that loads all configs from its local directory. **Is this expected behaviour, a bug or am I messing up the build process?**

2. A much bigger issue is that my Q8 build doesn't behave as a static installed official Q8 that I downloaded from mirrors. We have a very large and complex PSB file that Q16 can't "identify", but Q8 can, hence the whole endeavor to build a Q8 portable. If I install Q8 from one of the old mirrors, I can process the file, but I cannot with my own build of Q8 of the same version. Some differences in behaviour of my build comparing to the "official" one are:

- my build always thinks there is only 1GB of memory available unless I use `-limit` options to increase that. If I run `-list resource` I see memory capped at a little over 1GB even on a 32GB machine. However, if I run the official version of the same release I get over 7GB of memory without `-limit` options tweaks.
- when my build runs, it always runs at just over 1GB ram. The official build runs at whatever the limit is for this PSB file; no tweaks needed. According to procexp, the process allocates just over 1GB and never more; official dynamically gets more and more and it processes the file.
- my build fails on the large PSB file due to the "bad file descriptor error" (original error, where the journey to build started) without using `-limit` options, while official works fine. My build is only able to  run "identify.exe" on this large PSB file only if I increase `map` and `memory`, which is something I was observed with Q16.

So my questions are (in addition to the one highlighted above):
- Aside from running `identify.exe -version`, how can I verify that my build is indeed Q16? Any string I can search for in the binary or something like that?
- I understand this isn't simple to debug remotely. Can I provide any more details? Can anything in what I described be attributed to a specific mistake I am making during build?

Thank you in advance. Any help is appreciated!

**Comment by dlemstra** [comment]
Thanks for providing so much information. And I hope I can give you an answer to your questions.

1> I suspect that you are using the latest version of VisualMagick with an old version of ImageMagick? You should pick a commit of VisualMagick around the same time that 6.9.8-10 was released. But it might be possible that there are some bugs in that old version. 

2a> Would it be possible to share that large and complex PSB file through an email with a link to a location where we can download it? It is really strange that Q16 cannot read it but Q8 can. And have you tested if the latest version of ImageMagick can read that file? The reason that you see different behaviour is probably because you are creating a 32-bit version instead of a 64-bit versions. I would advise you to do 64-bit build. 

2b> Why would you want to use a different method than `identify.exe -version` to check if you have `Q8` or a `Q16` build?

**Reply by sergenikov** [reply]
Thank you for the reply Dirk; excited to get a reply from the community!

RE 1> I checked VisualMagick folder that's get's pulled when running `CloneRepositories.cmd` script from `ImageMagick6 -Windows` repo: `git remote -v` points to `https://github.com/ImageMagick/VisualMagick6.git`. It's on HEAD. I guess, that's not quite "latest" considering VM7, but also not the oldest. Would it be worth not using VisualMagick for this build and build without it?

RE 2a> Unfortunately, I cannot share the file due to company policy. I can tell it's a 6.2GB PSB file with approx 430 layers according to Photoshop with alphas and all the bells and whistles of complex layering; I don't know PS well, so can't say more. I also know size isn't the issue. We've processed a 9+GB files using IM and it handles them like a champ. This file is just complex. According to Photoshop it needs 18+GB of RAM to process. IM7 doesn't run on the file either with the same error (see below). The only ways I found to have identify.exe run on the file were:

1> On a machine with 32GB of ram and without any `-limit` configurations. Works every time and fast.
2> On a machine with even 4GB or ram, but with `-limit memory 2GiB -limit map 18GiB -limit disk 100GiB`, where map is the most important configuration. Dropping memory to 1GiB also works fine. Dropping map even below 18GiB results in the following error and this error has been haunting in my sleep me for a while:

```
UnableToWritePixelCache `D:/work/temp/magick-59765yTpO2QGNInd': Bad file descriptor @ error/cache.c/WritePixelCachePixels/5920
```

I don't think disk setting affects this at all and needs to be set to this high. I had a script count the total size of temp files and it's never over 10GB. Disk space is plentiful. Small disk space results in a different error specific to low disk space.

3> Run using Q8 (static, installed; downloaded from one of the old mirrors). Running with Q8 works on any machine without any `-limit` configurations. Since, Q8 portable isn't available, we need to build it.

The end result of running "identify.exe" is this line a large number of times with different resolutions.
```
Format : PSD Dimension : 6000x4897 IccProfile : Adobe RGB (1998) Colorspace : sRGB Opaque : true X Resolution: 300 Y Resolution: 300
```

RE 2b> The reason why I thought of trying getting application information other than `identify.exe -version` is that I don't trust the build. For example, configure.xml file remains with 6.9.10 versions in it even though I have 6.9.8-10 checked out by the tag. Variables inside it don't get expanded. This way, `identify.exe -version` shows 6.9.8-10, but `identify.exe -debug configure -list configure` shows 6.9.10-10 from ./configure.xml. So part of me isn't certain that it's building/packaging/bundling everything correctly especially not having luck building using configure.exe alone.

Can you suggest a way to build a portable 6.9.8-10 Q8 without using configure? Also, what other information can I provide you with about this for you to help out more? I've done a lot of testing and don't mind contributing somehow if you feel there may be a bug especially given IM7 doesn't work on this as well.

Thank you so much for trying to help, Dirk!

P.S. Also, I tried all suggestions in [massive file handling](https://www.imagemagick.org/Usage/files/#massive) with the same error.

**Comment by dlemstra** [comment, marked as answer]
We now have a portable version of the Q8 build in our downloads.
